src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
+            if (!_disposed)
+            {
+                // Running callers might still be using the cancellation token.
+                // Mark it canceled but don't dispose of the source while the callers are running.
+                // Otherwise, callers would receive ObjectDisposedException when calling token.Register.
+                // For now, rely on finalization to clean up _cancellationTokenSource's wait handle (if allocated).
+                _cancellationTokenSource.Cancel();
+
+                if (_receiver != null && _receiver.IsValueCreated)
+                {
+                    Receiver.CloseAsync().Wait();
+                    _receiver = null;
+                }
+
+                if (_clientEntity != null)
+                {
+                    _clientEntity.CloseAsync().Wait();
+                    _clientEntity = null;
+                }
+
+                _disposed = true;
+            }
+        }
+
         private Lazy<MessageReceiver> CreateMessageReceiver()
         {
             return new Lazy<MessageReceiver>(() => _messagingProvider.CreateMessageReceiver(_entityPath, _serviceBusAccount.ConnectionString));
-        }
-
-        internal async Task ProcessMessageAsync(Message message, CancellationToken cancellationToken)
-        {
-            if (!await _messageProcessor.BeginProcessingMessageAsync(message, cancellationToken))
-            {
-                return;
-            }
-
-            FunctionResult result = await _triggerExecutor.ExecuteAsync(message, cancellationToken);
-            await _messageProcessor.CompleteProcessingMessageAsync(message, result, cancellationToken);
-        }
-
-        internal async Task ProcessSessionMessageAsync(IMessageSession session, Message message, CancellationToken cancellationToken)
-        {
-            _messageSession = session;
-            if (!await _sessionMessageProcessor.BeginProcessingMessageAsync(session, message, cancellationToken))
-            {
-                return;
-            }
-
-            FunctionResult result = await _triggerExecutor.ExecuteAsync(message, cancellationToken);
-            await _sessionMessageProcessor.CompleteProcessingMessageAsync(session, message, result, cancellationToken);
-        }
-
-        private void ThrowIfDisposed()
-        {
-            if (_disposed)
-            {
-                throw new ObjectDisposedException(null);
-            }
-        }
-
+        }
+
+        internal async Task ProcessMessageAsync(Message message, CancellationToken cancellationToken)
+        {
+            if (!await _messageProcessor.BeginProcessingMessageAsync(message, cancellationToken))
+            {
+                return;
+            }
+
+            ServiceBusTriggerInput input = ServiceBusTriggerInput.New(message);
+            input.MessageReceiver = Receiver;
+
+            FunctionResult result = await _triggerExecutor.TryExecuteAsync(GetTriggerFunctionData(message, input), cancellationToken);

```

Comment:
nit: put the triggered function data on its own line, assigned to its own var (for this and other occurrences below).